src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs

```diff
@@ -299,6 +307,43 @@ static void AppendHashtableContents(IndentedTextWriter writer, Hashtable ht)
             }
         }
 
+        private static void EmitScan(IndentedTextWriter writer, RegexMethod rm, string id)
+        {
+            using (EmitBlock(writer, "while (true)"))
+            {
+                using (EmitBlock(writer, "if (FindFirstChar(text))"))
+                {
+                    if (rm.MatchTimeout != int.MaxValue)
+                    {
+                        writer.WriteLine("base.CheckTimeout();");
+                        writer.WriteLine();
+                    }
+
+                    writer.WriteLine("// If we got a match, we're done.");
+                    using (EmitBlock(writer, "if (Go(text))"))
```

Comment:
So the idea is that Go returns a bool in addition to calling Capture(0, ...) so that the details of the Match object and its layout remains private.  I like that, that makes sense.

Reply:
Oh, wait, this is a private Go helper. Still fine, but likely to go away once we eventually inline FindFirstChar and Go into here.  Regardless, I like now that the source gen'd Scan needs to know nothing about Match.

Reply:
right, the main reason why I went with this approach was because currently on the internal engines, after calling `Go` we check whether it actually called Capture(0) by checking `match._matchCount[0]` to be greater than 0, but we can't do that in the sourcegenerated version of Scan because that field is internal. Also, in order to be able to call something like `match.Success` we would need to call `match.Tidy(..)` first, which again would cause a small perf hit. So my thoughts were that withing the source generated type, we do know if Capture(0) was called or not in the private Go() method, so I decided to move that to a bool instead that notifies Scan wether a match was found or not.